src/serializer/ebml.rs

```diff
@@ -15,381 +20,245 @@ pub fn vint_size(i: u64) -> u8 {
+                Ok(next) => {
+                    input = next;
+                }
+                Err(e) => return Err(e),
+            }
 
-    let mut i = needed_bytes - 1;
-    loop {
-        match gen_be_u8!(input, (num.wrapping_shr((i * 8).into())) as u8) {
-            Ok(next) => {
-                input = next;
+            if i == 0 {
+                break;
             }
-            Err(e) => return Err(e),
+            i -= 1;
         }
 
-        if i == 0 {
-            break;
-        }
-        i -= 1;
+        Ok(input)
     }
-
-    Ok(input)
 }
 
 //FIXME: is it the right implementation?
-pub fn gen_int(mut input: (&mut [u8], usize), num: i64) -> Result<(&mut [u8], usize), GenError> {
-    let needed_bytes = vint_size(num as u64);
+pub(crate) fn gen_int(
+    num: i64,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError> {
+    move |mut input| {
+        let needed_bytes = vint_size(num as u64);
+
+        let mut i = needed_bytes - 1;
+        loop {
+            match set_be_i8((input.0, input.1), (num >> (i * 8)) as i8) {
+                Ok(next) => {
+                    input = next;
+                }
+                Err(e) => return Err(e),
+            }
 
-    let mut i = needed_bytes - 1;
-    loop {
-        match gen_be_i8!(input, (num >> (i * 8)) as i8) {
-            Ok(next) => {
-                input = next;
+            if i == 0 {
+                break;
             }
-            Err(e) => return Err(e),
+            i -= 1;
         }
 
-        if i == 0 {
-            break;
-        }
-        i -= 1;
+        Ok(input)
     }
-
-    Ok(input)
 }
 
-pub fn gen_u64(
-    input: (&mut [u8], usize),
+fn gen_type<T: Copy, G>(
     id: u64,
-    num: u64,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 8) >> gen_be_u64!(num)
-    )
+    size: u64,
+    num: T,
+    f: G,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError>
+where
+    G: Fn((&mut [u8], usize), T) -> Result<(&mut [u8], usize), GenError>,
+{
+    move |input| {
+        let temp = gen_vid(id)(input)?;
+        let temp = gen_vint(size)(temp)?;
+        f(temp, num)
+    }
 }
 
-pub fn gen_u32(
-    input: (&mut [u8], usize),
+pub(crate) fn gen_f64(
     id: u64,
-    num: u32,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 4) >> gen_be_u32!(num)
-    )
+    num: f64,
+) -> impl Fn((&mut [u8], usize)) -> Result<(&mut [u8], usize), GenError> {
+    move |input| gen_type(id, 8, num, set_be_f64)(input)
 }
 
-pub fn gen_u16(
-    input: (&mut [u8], usize),
+pub(crate) fn gen_f64_ref<'a>(
     id: u64,
-    num: u16,
-) -> Result<(&mut [u8], usize), GenError> {
-    do_gen!(
-        input,
-        gen_call!(gen_vid, id) >> gen_call!(gen_vint, 2) >> gen_be_u16!(num)
-    )
+    num: &'a f64,
```

Comment:
Also here the lifetime. You can avoid that by dereferencing outside the closure below